### PR TITLE
Introduce hasSession field in datasource

### PR DESCRIPTION
**Pull request by sgdc3**
That makes isLogged more consistent as it will be '1' only when the player is online.

TODO:
- Unit testing?

### CODE REVIEW

**Comment by sgdc3** [comment]
What do you think? My main concerns are performance and security.

File: docs/config.md [hunk 1]
```diff
@@ -40,6 +40,8 @@ DataSource:
     mySQLColumnEmail: 'email'
     # Column for storing if a player is logged in or not
     mySQLColumnLogged: 'isLogged'
+    # Column for storing if a player has a valid session or not
+    mySQLColumnHasSession: 'isLogged'
```
Comment (ljacqu) [inline docs/config.md]: Intentional to have the same name as `mySQLColumnLogged`? Maybe have an explanation for that?

File: src/main/java/fr/xephi/authme/command/executable/authme/RegisterAdminCommand.java [hunk 2]
```diff
@@ -51,38 +51,31 @@ public void executeCommand(final CommandSender sender, List<String> arguments) {
...
-                    commonService.send(sender, MessageKey.ERROR);
-                    return;
-                }
-                dataSource.setUnlogged(playerNameLowerCase);
+            if (!dataSource.saveAuth(auth)) {
+                commonService.send(sender, MessageKey.ERROR);
+                return;
+            }
+            dataSource.setUnlogged(playerNameLowerCase);
+            dataSource.revokeSession(playerNameLowerCase);
```
Comment (ljacqu) [inline src/main/java/fr/xephi/authme/command/executable/authme/RegisterAdminCommand.java]: are the above two calls necessary as a player gets saved? I'd expect those are the defaults.

File: src/main/java/fr/xephi/authme/datasource/MySQL.java [hunk 3]
```diff
@@ -566,6 +566,44 @@ public void setUnlogged(String user) {
     }
```
Comment (ljacqu) [inline src/main/java/fr/xephi/authme/datasource/MySQL.java]: Initialization of `hasSession` if missing should be added to `checkTablesAndColumns()`

File: src/main/java/fr/xephi/authme/datasource/SQLite.java [hunk 4]
```diff
@@ -467,6 +467,48 @@ public void setUnlogged(String user) {
     }
 
     @Override
+    public boolean hasSession(String user) {
+        String sql = "SELECT * FROM " + tableName + " WHERE LOWER(" + col.NAME + ")=?;";
```
Comment (ljacqu) [inline src/main/java/fr/xephi/authme/datasource/SQLite.java]: `SELECT col.HAS_SESSION` would be better as it retrieves only what we want from the DB.

File: src/main/java/fr/xephi/authme/datasource/SQLite.java [hunk 5]
```diff
@@ -467,6 +467,48 @@ public void setUnlogged(String user) {
     }
```
Comment (ljacqu) [inline src/main/java/fr/xephi/authme/datasource/SQLite.java]: Like in the MySQL class, here I'm missing the initialization of the `hasSession` column

File: src/main/java/fr/xephi/authme/datasource/SQLite.java [hunk 6]
```diff
@@ -467,6 +467,48 @@ public void setUnlogged(String user) {
...
+            pst.executeUpdate();
+        } catch (SQLException ex) {
+            logSqlException(ex);
+        }
+    }
+
+    @Override
+    public void revokeSession(String user) {
+        String sql = "UPDATE " + tableName + " SET " + col.HAS_SESSION + "=? WHERE LOWER(" + col.NAME + ")=?;";
+        if (user != null) {
```
Comment (ljacqu) [inline src/main/java/fr/xephi/authme/datasource/SQLite.java]: Why check here that `user != null` if this is not done in MySQL or in `SQLite#grantSession`? I don't think it's necessary.

File: src/main/java/fr/xephi/authme/process/join/AsynchronousJoin.java [hunk 7]
```diff
@@ -179,8 +179,9 @@ private void processJoinSync(Player player, boolean isAuthAvailable, Location lo
 
     private boolean canResumeSession(Player player) {
```
Comment (ljacqu) [inline src/main/java/fr/xephi/authme/process/join/AsynchronousJoin.java]: What would you think about introducing a `SessionService` that would encapsulate the session handling logic? This has become a large method that I don't really feel is a good fit inside `AsynchronousJoin`.

File: src/main/java/fr/xephi/authme/process/logout/AsynchronousLogout.java [hunk 8]
```diff
@@ -53,6 +53,7 @@ public void logout(Player player) {
 
         playerCache.removePlayer(name);
         database.setUnlogged(name);
+        database.revokeSession(name);
```
Comment (ljacqu) [inline src/main/java/fr/xephi/authme/process/logout/AsynchronousLogout.java]: It looks like `revokeSession` and `setUnlogged` are always called together. If we merge this into one call we would save some DB operations, which would be good.

File: src/main/java/fr/xephi/authme/process/quit/AsynchronousQuit.java [hunk 9]
```diff
@@ -82,8 +82,11 @@ public void processQuit(Player player) {
         playerCache.removePlayer(name);
 
         //always update the database when the player quit the game (if sessions are disabled)
-        if(!wasLoggedIn || !service.getProperty(PluginSettings.SESSIONS_ENABLED)) {
+        if(wasLoggedIn) {
             database.setUnlogged(name);
+            if(!service.getProperty(PluginSettings.SESSIONS_ENABLED)) {
```
Comment (ljacqu) [inline src/main/java/fr/xephi/authme/process/quit/AsynchronousQuit.java]: here and above, would be nice to have a space before reserved keywords like `if` `while` `for` etc.
Same for the `AsynchronousLogin` addition

File: src/main/java/fr/xephi/authme/process/join/AsynchronousJoin.java [hunk 10]
```diff
@@ -179,8 +179,9 @@ private void processJoinSync(Player player, boolean isAuthAvailable, Location lo
...
         final String name = player.getName();
-        if (database.isLogged(name)) {
+        if (database.hasSession(name)) {
             database.setUnlogged(name);
+            database.revokeSession(name);
             playerCache.removePlayer(name);
             if(service.getProperty(PluginSettings.SESSIONS_ENABLED)) {
```
Comment (ljacqu) [inline src/main/java/fr/xephi/authme/process/join/AsynchronousJoin.java]: `playerCache.removePlayer(name);` necessary? A player should never be in the PlayerCache at this point in time.

File: docs/config.md (hunk 1 above)
Comment (sgdc3) [reply docs/config.md]: LOL, i shoudn't code on the train.

File: src/main/java/fr/xephi/authme/command/executable/authme/RegisterAdminCommand.java (hunk 2 above)
Comment (sgdc3) [reply src/main/java/fr/xephi/authme/command/executable/authme/RegisterAdminCommand.java]: Right, good point

File: src/main/java/fr/xephi/authme/datasource/MySQL.java (hunk 3 above)
Comment (sgdc3) [reply src/main/java/fr/xephi/authme/datasource/MySQL.java]: Yeah, again, i shouldn't code when i'm on a crowded train xD

File: src/main/java/fr/xephi/authme/datasource/SQLite.java (hunk 4 above)
Comment (sgdc3) [reply src/main/java/fr/xephi/authme/datasource/SQLite.java]: Yeah, i left the previous code, i'll fix that asap

**Comment by sgdc3** [comment]
isLogged will be just a flag that can be used by website integrations

File: src/main/java/fr/xephi/authme/process/join/AsynchronousJoin.java (hunk 10 above)
Comment (sgdc3) [reply src/main/java/fr/xephi/authme/process/join/AsynchronousJoin.java]: Probably some code left there from the previous bungeecord cache implementation

File: src/main/java/fr/xephi/authme/process/logout/AsynchronousLogout.java (hunk 8 above)
Comment (sgdc3) [reply src/main/java/fr/xephi/authme/process/logout/AsynchronousLogout.java]: Nope, we don't do that when the player quit the server.

**Comment by ljacqu** [comment]
Hi @sgdc3, if there are no changes on your end and you're satisfied with my additions, I invite you to merge the PR. ;)

**Comment by ljacqu** [comment]
I took the liberty of squashing this PR so that we could have the issue number in the commit. This way, should we use Git blame months down the road to understand something, we can easily look up the issue for context.

**Comment by sgdc3** [comment]
@ljacqu Uh, sorry yesterday i wasn't at home and i didn't read your comment ;)

**Pull request merged**